src/runtime/runtime.ts

```diff
@@ -288,86 +336,357 @@ class Scan {
    * (i.e., this scan could not be satisfied due to proposals from previous scans) we'll return false.
    */
   applyInput(input:Change, prefix:ID[]) {
-    let resolved = this.resolve(prefix);
-
     // If this change isn't relevant to this scan, skip it.
-    if(this.notStaticMatch(input, "e")) return true;
-    if(this.notStaticMatch(input, "a")) return true;
-    if(this.notStaticMatch(input, "v")) return true;
-    if(this.notStaticMatch(input, "n")) return true;
+    if(this.notStaticMatch(input, "e")) return ApplyInputState.none;
+    if(this.notStaticMatch(input, "a")) return ApplyInputState.none;
+    if(this.notStaticMatch(input, "v")) return ApplyInputState.none;
+    if(this.notStaticMatch(input, "n")) return ApplyInputState.none;
 
     // For each register field of this scan, if the required value is impossible fail, otherwise add this new value to the
     // appropriate register in the prefix.
     // @NOTE: Technically, we republish existing values here too. In practice, that's harmless and eliminates the need for extra
     // branching.
     if(isRegister(this.e)) {
-      if(prefix[this.e.offset] !== undefined && prefix[this.e.offset] !== input.e) return false;
+      if(prefix[this.e.offset] !== undefined && prefix[this.e.offset] !== input.e) return ApplyInputState.fail;
       prefix[this.e.offset] = input.e;
     }
 
     if(isRegister(this.a)) {
-      if(prefix[this.a.offset] !== undefined && prefix[this.a.offset] !== input.a) return false;
+      if(prefix[this.a.offset] !== undefined && prefix[this.a.offset] !== input.a) return ApplyInputState.fail;
       prefix[this.a.offset] = input.a;
     }
 
     if(isRegister(this.v)) {
-      if(prefix[this.v.offset] !== undefined && prefix[this.v.offset] !== input.v) return false;
+      if(prefix[this.v.offset] !== undefined && prefix[this.v.offset] !== input.v) return ApplyInputState.fail;
       prefix[this.v.offset] = input.v;
     }
 
     if(isRegister(this.n)) {
-      if(prefix[this.n.offset] !== undefined && prefix[this.n.offset] !== input.n) return false;
+      if(prefix[this.n.offset] !== undefined && prefix[this.n.offset] !== input.n) return ApplyInputState.fail;
       prefix[this.n.offset] = input.n;
     }
 
+    return ApplyInputState.pass;
+  }
+
+  propose(index:Index, prefix:ID[], transaction:number, round:number, results:any[]):Proposal {
+    let {e,a,v,n} = this.resolve(prefix);
+    this.proposal.skip = false;
+    let proposal = index.propose(this.proposal, e, a, v, n, transaction, round);
+
+    let ix = 0;
+    for(let field of proposal.forFields) {
+      proposal.forRegisters[ix] = this[field] as Register;
```

Comment:
You spotted a very obnoxious bug that we ended up fixing in the future. Good eye! Luckily this one broke things badly enough that it didn't go unnoticed for long. Unfortunately in the pursuit of fast code ugly and tough to write things like this have to be done sometimes. If you see any other issues like this, please don't hesitate to point them out -- a spare set of eyes really makes fixing these much easier!